api/envoy/extensions/filters/http/api_key_auth/v3/api_key_auth.proto

```diff
@@ -28,27 +26,85 @@ option (xds.annotations.v3.file_status).work_in_progress = true;
-  // The clientID serves solely for identification purposes and isn't used for authentication.
-  config.core.v3.DataSource keys = 1 [(udpa.annotations.sensitive) = true];
+message ApiKeyAuth {
+  // The credentials that are used to authenticate the clients.
+  Credentials credentials = 1 [(udpa.annotations.sensitive) = true];
+
+  // The key sources to fetch the key from the coming request.
+  KeySources key_sources = 2;
+}
 
-  // The header name to fetch the key.
-  // If multiple values are present in the given header, the filter rejects the request.
-  // Only one of authentication_header, authentication_query, or authentication_cookie should be set.
-  string authentication_header = 2;
+// API key auth configuration of per route or per virtual host or per route configuration.
+message ApiKeyAuthPerRoute {
+  // Route specific APIKeyAuth configuration. This is optional and could be used to override the
+  // filter level ApiKeyAuth configuration **partly**.
+  //
+  // For example, if route specific ``authentication_header`` is set, then the filter will fetch
+  // the key from route specific ``authentication_header``. But the filter level ``keys`` will
+  // still be used. If route specific non-empty ``keys`` is set, then the filter will use the
+  // route specific ``keys`` to authenticate the client but the filter level ``authentication_header``
+  // will still be used. If both ``authentication_header`` and ``keys`` are set, then the filter
+  // will use the route specific ``authentication_header`` and ``keys`` to authenticate the client.
+  ApiKeyAuth override_config = 1;
 
-  // The query parameter name to fetch the key.
-  // Only one of authentication_header, authentication_query, or authentication_cookie should be set.
-  string authentication_query = 3;
+  // A list of client ids that are allowed to access the routes of the scope. If the list is empty,
+  // then all authenticated clients are allowed. This provides very limited but simple authorization.
+  // If more complex authorization is required, then use the :ref:`HTTP RBAC filter
+  // <config_http_filters_rbac>` instead.
+  repeated string allowed_clients = 2;
+}
+
+// Single credential entry that contains the API key and the related client id.
+message Credential {
+  // The value of the unique API key.
+  string key = 1 [(validate.rules).string = {min_len: 1}];
+
+  // The unique id or identity that used to identify the client or consumer.
+  string client_id = 2 [(validate.rules).string = {min_len: 1}];
+}
+
+// The credentials that are used to authenticate the clients.
+message Credentials {
+  // The list of credentials. Explicitly setting the Credentials message with zero entries is
+  // allowed and will result in the filter rejecting all requests.
+  repeated Credential entries = 1;
+}
+
+message KeySource {
+  // The header name to fetch the key. If multiple header values are present, the first one will be
+  // used.
```

Comment:
Please add a note about the "Bearer " prefix as it seems to be needed by the code.

Reply:
The prefix is not necessary. But if it present, we will stripped it.